### ISSUE TITLE
small additions for you

### ISSUE THREAD

**Issue by janmast** [arieh/ScrollBar#2]
Hi, I found to be in need options of 2 extra options when initiating a new ScrollerBar, so I was so free to add them :

example : 
var scroll = new ScrollerBar($('content-box-item'),{mode: 'vertical', margins: 20, pos: 'left', hide: true});

1\* option pos : allows me to let the scroller be on the left of right (for vertical) or top of bottom (for horizontal) scrollers AND attach a separate style class for each of these options.

2\* option hide : allows me to show or hide the handle if there is nothing to be scrolled anyway (ratio > 0)

changes : 

js :

...

```
Implements : [Options,Events]
, options :{
    step : 30
    , scrollerHtml : '<span class="decrease"></span>'
        +'<div class="scroll"><span class="handle"></span></div>'
        +'<span class="increase"></span>'
    , mode : 'vertical'
    , rtl : false
    , margins : 0
    , wrapped : null
    , pos : 'left'        // added
    , hide : false     // added
}
```

...

 var scroller = this.scroller.element = new Element('div',{"class":'scroller',html:this.options.scrollerHtml}).addClass(this.options.mode).addClass(this.options.pos)

// added extra class
...

this.areaSize =  this.element.getDimensions()[this.axis];
this.scrolledSize = this.scrolled.getDimensions()[this.axis];
this.scrollerSize = this.scroller.scroll.getDimensions()[this.axis];

// srollSize has been renamed to scrolledSize everywhere, scrollerSize has been added

ratio = this.areaSize / this.scrolledSize;

handleSize = (this.options.hide) ? (ratio >= 1) ? 0 : +this.scrollerSize \* ratio : (ratio >= 1) ? this.areaSize : +this.scrollerSize \* ratio;

// the previous version allowed a handle to be larger than the scrollable area : areasize: 400px / scrolledSize: 125px = ration of 3.2. making the handleSize : 400 x 3.2 = 1280 px !

full code : 

(function(window,$,undef){

"use strict";

var params = {
    Implements : [Options,Events]
    , options :{
        step : 30
        , scrollerHtml : '<span class="decrease"></span>'
            +'<div class="scroll"><span class="handle"></span></div>'
            +'<span class="increase"></span>'
        , mode : 'vertical'
        , rtl : false
        , margins : 0
        , wrapped : null
        , pos : 'left'  
        , hide : false
    }
    , element : null 
    , scroller : null
    , scrolledSize : 0 
    , areaSize : 0
    , position: 0
    , slider: null
    , events : null
    , generated : false
    , attached : false
    , axis : 'y'
    , dir : 'top'
    , property : 'height'
    , initialize : function initialize(elem,opts){
        this.setOptions(opts);
        this.element = $(elem);

```
    this.axis = (this.options.mode == 'vertical') ? 'y' : 'x';
    this.dir = (this.options.mode == 'vertical') ? 'top' : this.options.rtl ? 'right' : 'left';
    this.property = (this.options.mode == 'vertical') ? 'height' : 'width';

    this.constrcut();
    this.attach();
}
, constrcut : function constrcut(){
    this.scroller = {};

    var scroller = this.scroller.element = new Element('div',{"class":'scroller',html:this.options.scrollerHtml}).addClass(this.options.mode).addClass(this.options.pos)
        , ratio
        , $this = this
        , handleSize;

    this.scroller.inc = scroller.getElement('.decrease');
    this.scroller.dec  = scroller.getElement('.increase');
    this.scroller.scroll = scroller.getElement('.scroll');
    this.scroller.handle = scroller.getElement('.handle');

    this.scrolled = (this.options.wrapped) ? this.options.wrapped : (function(){
        var html = $this.element.get('html');
        $this.element.empty();
        return new Element('div',{html:html,'class':'wrapped'}).inject($this.element);
    }());

    this.element.setStyle('overflow','hidden');

    this.element.adopt(scroller);

    this.areaSize =  this.element.getDimensions()[this.axis];
    this.scrolledSize = this.scrolled.getDimensions()[this.axis];
    this.scrollerSize = this.scroller.scroll.getDimensions()[this.axis];

    ratio = this.areaSize / this.scrolledSize;

    handleSize = (this.options.hide) ? (ratio >= 1) ? 0 : +this.scrollerSize * ratio : (ratio >= 1) ? this.areaSize : +this.scrollerSize * ratio;

    this.scroller.handle.setStyle(this.property,handleSize);

    this.slider = new Slider(this.scroller.scroll,this.scroller.handle,{mode:this.options.mode, range : [0,this.scrolledSize-this.areaSize/2+this.options.margins]});

    this.generated = true;
}
, attach : function attach(){
    var $this = this;

    if (this.attached) return;

    if (!this.generated) this.generate();

    this.events = {
        scrollUp :function scrollUp(){
            $this.slider.set($this.position - $this.options.step);
        }
        , scrollDown : function scrollDown(){
            $this.slider.set($this.position + $this.options.step);
        }
        , manageWheel : function manageWheel(e){
            e.preventDefault();
            if (e.wheel >0){
                $this.events.scrollUp();
            }else{
                $this.events.scrollDown();
            }    
        }
    };

    this.element.addEvent('mousewheel',this.events.manageWheel);

    this.scroller.inc.addEvent('click',this.events.scrollUp);
    this.scroller.dec.addEvent('click',this.events.scrollDown);

    this.slider.addEvent('change' , function(pos){
        if (pos < $this.position) $this.decrease($this.position - pos);
        else $this.increase(pos - $this.position);
    });

    this.attached = true;
}
, detach : function detach(){
    this.element.removeEvent('mousewheel',this.events.manageWheel);
    this.slider.detach();
    this.scroller.each(function(el){el.destroy();});
    this.generated = false;
    this.attached = false;
}
, increase : function increase(step){
    step = step || this.options.step;

    if (this.position + step > this.scrolledSize-this.areaSize/2) this.position = this.scrolledSize-this.areaSize/2;
    else this.position += step;

    this.scrolled.setStyle('margin-'+this.dir,-1*this.position);
    this.fireEvent('increase',[this.position]);
}
, decrease : function decrease(step){
    step = step || this.options.step;

    if (this.position-step < 0) this.position = 0;
    else this.position -=step;

    this.scrolled.setStyle('margin-'+this.dir,-1*this.position);
    this.fireEvent('decrease',[this.position]);
}
, toElement : function toElement(){return this.element;}
```

}, 
ScrollerBar = this.ScrollerBar = new Class(params);

}).apply(this,[this,document.id]);

**Comment by arieh** [comment]
can you make this into a pull request?

**Comment by janmast** [comment]
hi, working on a deadline now. I will soon. svn guy, so I nee dto install Git first. found a bug in the changes though.
- renamed the 'pos' option to 'style' as I found that I was using way more different styles for different types of scrolls on the same page.

Implements: [Options, Events],
        options: {
            step: 30,
            scrollerHtml: '<span class="decrease"></span>' + '<div class="scroll"><span class="handle"></span></div>' + '<span class="increase"></span>',
            mode: 'vertical',
            rtl: false,
            margins: 0,
            wrapped: null,
            style: null,
            hide: false
        }
- found that the this.slider = new Slider etc calculated an impossible range 
  example : 400 - 400/2 + margins = 200 + margins even if there was nothing to scroll.
  now it takes the range according to the ratio and using the handleBar calculated
  
  ```
      this.slider = (ratio >= 1) ? new Slider(this.scroller.scroll, this.scroller.handle, {
          mode: this.options.mode,
          range: [0, handleSize]
      }) : new Slider(this.scroller.scroll, this.scroller.handle, {
          mode: this.options.mode,
          range: [0, this.scrolledSize - this.areaSize / 2 + this.options.margins]
      });
  ```

**Comment by arieh** [comment]
well, for right/left position you got the rtl flag - but in general - I try as hard as posible to put all css in the css file 
I didn't go through your code but I would strongly advise you check if what you're trying to do should in fact be in the css file and not on the JS

as for the hide - I'll look into it - I had some trouble finding out the exact size of the scroll box on some browsers so it might be hard to calculate it properly

**Comment by janmast** [comment]
hi Arieh, 

great scrollerbar. thanks.

the rtl flag works fine to position the scroller on the left or right but it does not allow me to assign different css rules to different scrollerbars on the same page. for example : I have 3 scrollerbars beside each other, the left one is rtl:left, has a special coloured css and special positioning, the center one has rtl:right and his own special styling, the right one has rtl:left and an even different styling as it only appears in special conditions.

.scroller.vertical.right1 {
    height:100%;
    width:8px;
    top:10px;
    right:10px;
...
}

.scroller.vertical.left1 {
    height:100%;
    width:8px;
    top:0px;
    left:0px;
...
}

.scroller.vertical.left2 {
    height:100%;
    width:20px;
    top:30px;
...
}
now, as I am a fan of splitting code and design, I wanted to be able to declare a new ScrollerBar with its specific css styling attached. that's why the class flag appeared.

the hide flag works fine here now, on different browsers and platforms. 

cheers,
Jan

**Comment by arieh** [comment]
You are going on this backwards-
.mydiv1 .scroller .left{special}
.mydiv2 .scroller .left{special}

no need to touch the markup if all you do is change locations

**Comment by janmast** [comment]
Ok, I feel like I can't get myself explained properly...

> use the demo page.
> add four more divs, give them different ids
> declare 3 ScrollerBar objects using 
> var vert_scroll_1 = new ScrollerBar ($('vert_1'),{mode:'vertical',rtl:'right'});
> declare 3 ScrollerBar objects using 
> var vert_scroll_1 = new ScrollerBar ($('vert_1'),{mode:'vertical',rtl:'left'});
> make them all have a differnet sized handle (rounded, square, transparent, color,width) and margin towards the scrolling text
> please explain how you can do it currently, so I can remove my obsolete stuff ...

greets

**Comment by arieh** [comment]
here's an example:
http://tests.arieh.co.il/scroller.html

notice that the identification is done by parent not the child. I've just checked and rtl isn't even being used - I've apparently deprecated it. It's all in the css - and that's the entire point.

**Comment by janmast** [comment]
hi ...

I use a DIV with an ID that gets different types of content through ajax. I now added a nested DIV with an ID related to its content, so they call their specifically styled ScrollerBar when the data got in. so, I removed the style tag and got some sleep.

thanks again,
Jan
